src/cargo/ops/cargo_rustc/context.rs

```diff
@@ -1082,24 +1070,55 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
     }
 
     pub fn incremental_args(&self, unit: &Unit) -> CargoResult<Vec<String>> {
-        if self.incremental_enabled {
-            if unit.pkg.package_id().source_id().is_path() {
-                // Only enable incremental compilation for sources the user can modify.
-                // For things that change infrequently, non-incremental builds yield
-                // better performance.
-                // (see also https://github.com/rust-lang/cargo/issues/3972)
-                return Ok(vec![format!("-Zincremental={}",
-                                       self.layout(unit.kind).incremental().display())]);
-            } else if unit.profile.codegen_units.is_none() {
-                // For non-incremental builds we set a higher number of
-                // codegen units so we get faster compiles. It's OK to do
-                // so because the user has already opted into slower
-                // runtime code by setting CARGO_INCREMENTAL.
-                return Ok(vec![format!("-Ccodegen-units={}", ::num_cpus::get())]);
+        // There's a number of ways to configure incremental compilation right
+        // now. In order of descending priority (first is highest priority) we
+        // have:
+        //
+        // * `CARGO_INCREMENTAL` - this is blanket used unconditionally to turn
+        //   on/off incremental compilation for any cargo subcommand. We'll
+        //   respect this if set.
+        // * `profile.dev.incremental` - in `Cargo.toml` specific profiles can
+        //   be configured to enable/disable incremental compilation. This can
+        //   be primarily used to disable incremental when buggy for a project.
+        // * `build.incremental` - in `.cargo/config` this blanket key can
+        //   globally for a system configure whether incremental compilation is
+        //   enabled. Note that setting this to `true` will not actually affect
+        //   all builds though. For example a `true` value doesn't enable
+        //   release incremental builds, only dev incremental builds. This can
+        //   be useful to globally disable incremental compilation like
+        //   `CARGO_INCREMENTAL`.
```

Comment:
Hm, so the value `true` for `build.incremental` does not change anything, because `true && default` is just `default`. 

So, the only usage for this flag is to **disable** incremental compilation. So perhaps we can rename the flag to `build.disable_incremental` and allow only `true` value for it? 

Should this setting override even explicit profile incremental setting? If the primary use-case here is to work-around the possible bugs in incremental compilation, then I think a reliable way to turn off **all** incremental would be more valuable.

Reply:
Yeah this was mostly just meant to mirror `CARGO_INCREMENTAL` as a "global configuration option" to disable it everywhere (I don't think enabling it everywhere makes much sense for release profiles for example).

I personally prefer `incremental` at least though to `disable-incremental` in that it's a bit more clear (even if `incremental = true` doesn't actually do anything).

I think that makes sense as well yeah for .cargo/config to override profiles, I'll tweak that